src/MOI_wrapper.jl

```diff
@@ -1,58 +1,99 @@
-    (LQOI.VecLin, MOI.Nonnegatives),
-    (LQOI.VecLin, MOI.Nonpositives),
-    (LQOI.VecLin, MOI.Zeros)
-]
+mutable struct ConstraintInfo
+    row::Int
+    set::MOI.AbstractSet
+    # Storage for constraint names. Where possible, these are also stored in the
+    # Gurobi model.
+    name::String
+    ConstraintInfo(row::Int, set) = new(row, set, "")
+end
 
-mutable struct Optimizer <: LQOI.LinQuadOptimizer
-    LQOI.@LinQuadOptimizerBase(Model)
+mutable struct Optimizer <: MOI.ModelLike
+    # The low-level Gurobi model.
+    inner::Model
+    # The Gurobi environment. If `nothing`, a new environment will be created
+    # on `MOI.empty!`.
     env::Union{Nothing, Env}
+    # The current user-provided parameters for the model.
     params::Dict{String, Any}
-    # The next two fields are used to cleverly manage calls to `update_model!`.
+
+    # The next field is used to cleverly manage calls to `update_model!`.
     # `needs_update` is used to record whether an update should be called before
-    # accessing a model attribute (such as the value of a RHS term). One of the
-    # main pain-points when using Gurobi through JuMP is the need to access the
-    # number of variables. Unfortunately, in order to use the Gurobi API (i.e.
-    # `num_vars(model.inner)`), we need to call `update_model!`. To avoid
-    # frequent calls just for this case, we cache the number of variables in
-    # `num_variables`. There are calls in `add_variables` and `delete_variables`
-    # to adjust this value. In addition, we also cache the number of linear and
-    # quadratic constraints.
+    # accessing a model attribute (such as the value of a RHS term).
     needs_update::Bool
-    num_variables::Int
-    num_linear_constraints::Int
-    num_quadratic_constraints::Int
+
+    # A flag to keep track of MOI.Silent, which over-rides the OutputFlag
+    # parameter.
+    silent::Bool
+
+    # An enum to remember what objective is currently stored in the model.
+    objective_type::ObjectiveType
+
+    # A flag to keep track of MOI.FEASIBILITY_SENSE, since Gurobi only stores
+    # MIN_SENSE or MAX_SENSE. This allows us to differentiate between MIN_SENSE
+    # and FEASIBILITY_SENSE.
+    is_feasibility::Bool
+
+    # A mapping from the MOI.VariableIndex to the Gurobi column. VariableInfo
+    # also stores some additional fields like what bounds have been added, the
+    # variable type, and the names of SingleVariable-in-Set constraints.
+    variable_info::CleverDicts.CleverDict{MOI.VariableIndex, VariableInfo}
+
+    # An index that is incremented for each new constraint (regardless of type).
+    # We can check if a constraint is valid by checking if it is in the correct
+    # xxx_constraint_info. We should _not_ reset this to zero, since then new
+    # constraints cannot be distinguished from previously created ones.
+    last_constraint_index::Int
+    # ScalarAffineFunction{Float64}-in-Set storage.
+    affine_constraint_info::Dict{Int, ConstraintInfo}
```

Comment:
Why these are not Clever ones?

Reply:
Because the `last_constraint_index` counter is shared over the types of constraints, so that there would be large holes in the CleverDict Array automatically.

Note, that this is not necessary, since the `i` in `ConstraintIndex{F,S}(i)` must only be unique per `F`, `S` type. It would probably be better to switch to `last_{affine,sos,quadratic}_constraint_index` and `CleverDict` for each of them.

Reply:
Mainly because it's not performance critical. The only time we really need it is querying duals/ConstraintFunctions/etc. The extra hash doesn't matter. In contrast, we get column indices all the time, so it makes sense to make these fast.